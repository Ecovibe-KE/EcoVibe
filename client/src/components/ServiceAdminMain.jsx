import { Col } from "react-bootstrap"

function ServiceAdminMain({ serviceImage, serviceName, serviceDescription, serviceDuration, serviceCurrency, servicePrice, serviceStatus }) {

    function determineVisibility() {
        if (serviceStatus.toLowerCase() == "active") {
            return ""
        } else {
            return "invisible"
        }
    }

    return (
        <>
            <Col md="4">
                <div className="card rounded-2 shadow h-100">
                    <img alt="service image" className="img-fluid rounded-top-2 fixed-card-img" src={serviceImage} />
                    <div className="card-body text-start d-flex flex-column justify-content-evenly">
                        <h5>{serviceName}</h5>
                        <p className="mt-2">{serviceDescription}</p>
                        <hr />
                        <section className="d-flex justify-content-between align-items-center">
                            <p className="primary-color fw-bold m-0">{serviceDuration}</p>
                            <p className="primary-color fw-bold m-0">{serviceCurrency}{servicePrice}</p>
                            <p className={`active-status-color fw-bold border rounded-pill p-2 m-0 ${determineVisibility()}`} >{serviceStatus}</p>
                        </section>
                    </div>
                </div>
            </Col>
        </>
    )
}

export default ServiceAdminMain