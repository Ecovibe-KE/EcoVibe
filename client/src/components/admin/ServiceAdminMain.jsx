/* eslint-disable react-refresh/only-export-components */
import { Col } from "react-bootstrap";

// Separates duration into hours and minutes
const separateDuration = (str) => {
  // 1. hrs? – matches "hr" or "hrs"
  // 2. (?: … )? – makes the entire minutes group optional
  const pattern = /(?:(?<h>\d+)\s*hrs?)?\s*(?:(?<m>\d+)\s*min)?/i;
  const groups = str.match(pattern)?.groups || {};

  const hours = parseInt(groups.h || "0", 10);
  const minutes = parseInt(groups.m || "0", 10);

  return { hours, minutes };
};

export function displayDuration(duration) {
  const { hours, minutes } = separateDuration(duration);
  if (minutes < 1) {
    return `${hours} hr`;
  } else if (hours < 1) {
    return `${minutes} min`;
  } else {
    return `${hours} hr ${minutes} min`;
  }
}

function ServiceAdminMain({
  serviceId,
  serviceImage,
  serviceTitle,
  serviceDescription,
  serviceDuration,
  priceCurrency,
  servicePrice,
  serviceStatus,
  handleShowEdit,
  getServiceId,
  setFormData,
  setPreviewUrl,
  setOriginalServiceData,
  handleShowDelete,
}) {
  function determineVisibility() {
    if (serviceStatus.toLowerCase() === "active") {
      return "";
    } else {
      return "invisible";
    }
  }

  // Purpose
  // - display editServiceModal
  // - serviceId of clicked service
  // - populate formdata with clicked service data
  // - update clicked originalServiceData
  function handleEditClick() {
    handleShowEdit();
    getServiceId(() => serviceId);
    setFormData({
      serviceTitle: serviceTitle,
      serviceDescription: serviceDescription,
      priceCurrency: priceCurrency,
      servicePrice: servicePrice,
      serviceDuration: separateDuration(serviceDuration),
      serviceImage: serviceImage,
      serviceStatus: serviceStatus,
    });
    setOriginalServiceData({
      title: serviceTitle,
      description: serviceDescription,
      currency: priceCurrency,
      price: servicePrice,
      duration: serviceDuration,
      image: serviceImage,
      status: serviceStatus,
    });
    setPreviewUrl(() => serviceImage);
  }

  function handleDeleteClick() {
    handleShowDelete();
    getServiceId(() => serviceId);
    setOriginalServiceData({
      title: serviceTitle,
    });
  }

  return (
    <>
      <Col md="4">
        <div className="card rounded-2 shadow h-100">
          <div className="image-wrapper">
            <img
              alt="service image"
              src={serviceImage}
              className="img-fluid rounded-top-2"
            />

            {/* Overlay with buttons (hidden by default) */}
            <div className="hover-overlay">
              <button
                className="btn btn-warning btn-sm mx-1"
                onClick={handleEditClick}
              >
                Edit
              </button>
              <button
                className="btn btn-danger btn-sm mx-1"
                onClick={handleDeleteClick}
              >
                Delete
              </button>
            </div>
          </div>

          <div className="card-body text-start d-flex flex-column justify-content-evenly">
            <h5>{serviceTitle}</h5>
            <p className="mt-2">{serviceDescription}</p>
            <hr />
            <section className="d-flex justify-content-between align-items-center">
              <p className="primary-color fw-bold m-0">
                {displayDuration(serviceDuration)}
              </p>
              <p className="primary-color fw-bold m-0">
                {priceCurrency + " " + servicePrice}
              </p>
              <p
                className={`status-color fw-bold border rounded-pill p-2 m-0 ${determineVisibility()}`}
              >
                {serviceStatus}
              </p>
            </section>
          </div>
        </div>
      </Col>
    </>
  );
}

export default ServiceAdminMain;
