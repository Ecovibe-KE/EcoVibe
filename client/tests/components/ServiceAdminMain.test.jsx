import { render, screen, fireEvent } from "@testing-library/react";
import ServiceAdminMain, { displayDuration } from "../../src/components/admin/ServiceAdminMain";
import "@testing-library/jest-dom";

describe("ServiceAdminMain Component", () => {
    const mockProps = {
        serviceId: 1,
        serviceImage: "test-image.png",
        serviceTitle: "Eco Consulting",
        serviceDescription: "Helping companies become greener",
        serviceDuration: "2 hr 30 min",
        priceCurrency: "KES",
        servicePrice: 5000,
        serviceStatus: "active",
        handleShowEdit: vi.fn(),
        getServiceId: vi.fn(),
        setFormData: vi.fn(),
        setPreviewUrl: vi.fn(),
        setOriginalServiceData: vi.fn(),
        handleShowDelete: vi.fn(),
    };

    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders service details correctly", () => {
        render(<ServiceAdminMain {...mockProps} />);
        expect(screen.getByText("Eco Consulting")).toBeInTheDocument();
        expect(screen.getByText("Helping companies become greener")).toBeInTheDocument();
        expect(screen.getByText(/KES 5000/)).toBeInTheDocument();
        expect(screen.getByText(/active/i)).toBeInTheDocument();
    });

    it("renders duration correctly using displayDuration()", () => {
        expect(displayDuration("2 hr 30 min")).toBe("2 hr 30 min");
        expect(displayDuration("2 hr 0 min")).toBe("2 hr");
        expect(displayDuration("0 hr 45 min")).toBe("45 min");
        expect(displayDuration("0 hr 0 min")).toBe("0 hr");
    });

    it("calls handleShowEdit and updates props correctly on Edit click", () => {
        render(<ServiceAdminMain {...mockProps} />);
        fireEvent.click(screen.getByText("Edit"));
        expect(mockProps.handleShowEdit).toHaveBeenCalled();
        expect(mockProps.getServiceId).toHaveBeenCalled();
        expect(mockProps.setFormData).toHaveBeenCalledWith(
            expect.objectContaining({
                serviceTitle: "Eco Consulting",
                serviceDescription: "Helping companies become greener",
                priceCurrency: "KES",
                servicePrice: 5000,
            })
        );
        expect(mockProps.setOriginalServiceData).toHaveBeenCalled();
        expect(mockProps.setPreviewUrl).toHaveBeenCalled();
    });

    it("calls handleShowDelete and updates props correctly on Delete click", () => {
        render(<ServiceAdminMain {...mockProps} />);
        fireEvent.click(screen.getByText("Delete"));
        expect(mockProps.handleShowDelete).toHaveBeenCalled();
        expect(mockProps.getServiceId).toHaveBeenCalled();
        expect(mockProps.setOriginalServiceData).toHaveBeenCalledWith({
            title: "Eco Consulting",
        });
    });

    it("renders invisible class when serviceStatus is inactive", () => {
        render(<ServiceAdminMain {...mockProps} serviceStatus="inactive" />);
        const statusEl = screen.getByText(/inactive/i);
        expect(statusEl).toHaveClass("invisible");
    });

    it("handles missing image gracefully", () => {
        render(<ServiceAdminMain {...mockProps} serviceImage={null} />);
        const img = screen.getByAltText("service image");
        expect(img).toBeInTheDocument();
        expect(img.getAttribute("src")).toBeNull(); // React omits src attribute when null
    });

    it("renders proper HTML structure", () => {
        render(<ServiceAdminMain {...mockProps} />);
        const img = screen.getByAltText("service image");
        expect(img).toHaveClass("img-fluid", "rounded-top-2");
        expect(screen.getByRole("button", { name: /edit/i })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: /delete/i })).toBeInTheDocument();
    });

    // Edge case: duration regex weird formats
    it("handles irregular duration strings gracefully", () => {
        expect(displayDuration("3 hrs")).toBe("3 hr");
        expect(displayDuration("15 min")).toBe("15 min");
        expect(displayDuration("")).toBe("0 hr");
    });
});
