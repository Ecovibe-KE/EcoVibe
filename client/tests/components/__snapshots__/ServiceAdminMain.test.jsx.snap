// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceAdminMain Component > renders the service details correctly 1`] = `
<div>
  <div
    class="col-md-4"
  >
    <div
      class="card rounded-2 shadow h-100"
    >
      <div
        class="image-wrapper"
      >
        <img
          alt="service image"
          class="img-fluid rounded-top-2"
          src="https://example.com/image.png"
        />
        <div
          class="hover-overlay"
        >
          <button
            class="btn btn-warning btn-sm mx-1"
          >
            Edit
          </button>
          <button
            class="btn btn-danger btn-sm mx-1"
          >
            Delete
          </button>
        </div>
      </div>
      <div
        class="card-body text-start d-flex flex-column justify-content-evenly"
      >
        <h5>
          Eco Consultation
        </h5>
        <p
          class="mt-2"
        >
          We offer eco-friendly business solutions.
        </p>
        <hr />
        <section
          class="d-flex justify-content-between align-items-center"
        >
          <p
            class="primary-color fw-bold m-0"
          >
            2 hr 30 min
          </p>
          <p
            class="primary-color fw-bold m-0"
          >
            KES 5000
          </p>
          <p
            class="status-color fw-bold border rounded-pill p-2 m-0 "
          >
            active
          </p>
        </section>
      </div>
    </div>
  </div>
</div>
`;
